src/client/executor.rs

```
@@ -309,167 +285,152 @@ impl Client {
+                        retry = Some(ExecutionRetry {
+                            prior_txn_number: None,
+                            first_error: err,
+                        });
+                        continue;
+                    } else {
+                        return Err(err);
                     }
                 }
+            };
 
-                self.inner
-                    .topology
-                    .handle_application_error(
-                        server.address.clone(),
-                        err.clone(),
-                        HandshakePhase::after_completion(&conn),
-                    )
-                    .await;
-                // release the connection to be processed by the connection pool
-                drop(conn);
-                // release the selected server to decrement its operation count
-                drop(server);
-
-                if retryability == Retryability::Read && err.is_read_retryable()
-                    || retryability == Retryability::Write && err.is_write_retryable()
-                {
-                    self.execute_retry(&mut op, &mut session, txn_number, err)
-                        .await
-                } else {
-                    Err(err)
-                }
+            if session.is_none() {
+                implicit_session = self.start_implicit_session(&op).await?;
+                session = implicit_session.as_mut();
             }
-        }
-    }
 
-    async fn execute_retry<T: Operation>(
-        &self,
-        op: &mut T,
-        session: &mut Option<&mut ClientSession>,
-        prior_txn_number: Option<i64>,
-        first_error: Error,
-    ) -> Result<ExecutionOutput<T>> {
-        op.update_for_retry();
-
-        let server = match self.select_server(op.selection_criteria()).await {
-            Ok(server) => server,
-            Err(_) => {
-                return Err(first_error);
+            let retryability = self.get_retryability(&conn, &op, &session)?;
```

Is there any chance this could fail during a retry? If so, I think we should return the original error rather than the one that occurs in this call.

---

The only error this can return is an internal one indicating an impossible/invalid connection state, which I think would be valuable to expose if it does happen :)